File: euler/calculus.py
from typing import Iterator, Union
from math import isqrt, log, sqrt
from euler.big_int import BigInt


def sum_arithmetic_series(a: Union[int, float], d: Union[int, float], n: int) -> float:
    """Get sum of first `n` terms of arithmetic series

    Args:
        a: First number in series
        d: Common difference of series
        n: Number of terms in series

    Examples:
        >>> print(sum_arithmetic_series(1, 2, 3))
        9.0
        >>> print(sum_arithmetic_series(0.5, 0.75, 3))
        3.75
    """
    return n * (2 * a + d * (n - 1)) / 2


def sum_geometric_series(a: Union[int, float], r: Union[int, float], n: int) -> float:
    """Get sum of `n` first terms of geometric series

    Args:
        a: First number in series
        r: Common ratio of series
        n: Number of terms in series

    Examples:
        >>> print(sum_geometric_series(1, 2, 3))
        7.0
        >>> print(sum_geometric_series(0.5, 0.75, 3))
        1.15625
    """
    return a * (r ** n - 1) / (r - 1)


def fibonacci_numbers(n: int, t: type = int) -> Iterator[Union[int, BigInt]]:
    """Generate `n` first Fibonacci numbers `OEIS A000045 <https://oeis.org/A000045>`_

    Args:
        n: Number of Fibonacci numbers
        t: Type of numbers

    Yields:
        The next Fibonacci number in the range of 0 to `n`

    Examples:
        >>> print(list(fibonacci_numbers(10)))
        [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55]
        >>> print(list(fibonacci_numbers(10, BigInt)))
        [0, 1, 1, 2, 3, 5, 8, 13, 21, 34, 55]
    """
    f_previous, f_current = t(0), t(1)
    yield f_previous
    yield f_current
    for i in range(2, n + 1):
        f_previous, f_current = f_current, f_previous + f_current
        yield f_current


def get_fibonacci(n: int) -> int:
    """Get `n`-th Fibonacci number

    Args:
        n: Index of Fibonacci number

    Examples:
        >>> print(get_fibonacci(10))
        55
    """
    return int(((1 + sqrt(5)) / 2) ** n / sqrt(5) + 0.5)


def fibonacci_index(n: int) -> int:
    """Get index of `n` in Fibonacci sequence

    Returns:
        Index of Fibonacci number n. If n isn't Fibonacci number, returns index of closest Fibonacci number below n

    Examples:
        `n` is Fibonacci number

        >>> print(fibonacci_index(55))
        10

        `n` isn't Fibonacci number

        >>> print(fibonacci_index(54))
        9
    """
    return int(log(n * sqrt(5) + 0.5, (sqrt(5) + 1) / 2))


def len_collatz_chain(n: int, hashmap: dict[int, int]) -> int:
    """Get length of Collatz chain

    Args:
        n: Starting number
        hashmap: Collatz chains lengths

    Examples:
        >>> print(len_collatz_chain(13, {1: 1}))
        10

    Notes:
        The functions uses memoization, so hashmap is required.
    """
    if n not in hashmap:
        if n % 2:
            hashmap[n] = 2 + len_collatz_chain((3 * n + 1) // 2, hashmap)
        else:
            hashmap[n] = 1 + len_collatz_chain(n // 2, hashmap)
    return hashmap[n]


def triangular_numbers(n: int) -> Iterator[int]:
    """Generate `n` first numbers in triangular numbers sequence `OEIS A000217 <https://oeis.org/A000217>`_

    Args:
        n: Number of triangular numbers

    Yields:
        The next triangular number in the range of 0 to `n`

    Examples:
        >>> print(list(triangular_numbers(10)))
        [0, 1, 3, 6, 10, 15, 21, 28, 36, 45, 55]
    """
    t = 0
    yield t
    for i in range(1, n + 1):
        t += i
        yield t


def get_triangular(n: int) -> int:
    """Get `n`-th triangular number

    Args:
        n: Index of triangular number

    Examples:
        >>> print(get_triangular(10))
        55
    """
    return n * (n + 1) // 2


def triangular_index(n: int) -> int:
    """Get index of `n` in triangular sequence

    Returns:
        Index of triangular number n. If n isn't triangular number, returns index of closest triangular number below n

    Examples:
        `n` is triangular number

        >>> print(triangular_index(55))
        10

        `n` isn't triangular number

        >>> print(triangular_index(54))
        9
    """
    return (isqrt(8 * n + 1) - 1) // 2


def is_triangular(n: int) -> bool:
    """Check if number is triangular

    Returns:
        True if number is triangular, False otherwise

    Examples:
        >>> is_triangular(55)
        True
        >>> is_triangular(54)
        False
    """
    return sqrt(8 * n + 1) % 2 == 1


def square_numbers(n: int) -> Iterator[int]:
    """Generate `n` first numbers in square numbers sequence `OEIS A000290 <https://oeis.org/A000290>`_

    Args:
        n: Number of square numbers

    Yields:
        The next square number in the range of 0 to `n`

    Examples:
        >>> print(list(square_numbers(10)))
        [0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]
    """
    t = 0
    yield t
    for i in range(1, 2 * n + 1, 2):
        t += i
        yield t


def get_square(n: int) -> int:
    """Get `n`-th square number

    Args:
        n: Index of square number

    Examples:
        >>> print(get_square(100))
        100
    """
    return n * n


def square_index(n: int) -> int:
    """Get index of `n` in square sequence

    Returns:
        Index of square number n. If n isn't square number, returns index of closest square number below n

    Examples:
        `n` is square number

        >>> print(square_index(100))
        10

        `n` isn't square number

        >>> print(triangular_index(99))
        9
    """
    return isqrt(n)


def is_square(n: int) -> bool:
    """Check if number is square

    Returns:
        True if number is square, False otherwise

    Examples:
        >>> is_square(55)
        True
        >>> is_square(54)
        False
    """
    return sqrt(n) % 1 == 0


def pentagonal_numbers(n: int) -> Iterator[int]:
    """Generate `n` first numbers in pentagonal numbers sequence `OEIS A000326 <https://oeis.org/A000326>`_

    Args:
        n: Number of pentagonal numbers

    Yields:
        The next pentagonal number in the range of 0 to `n`

    Examples:
        >>> print(list(pentagonal_numbers(10)))
        [0, 1, 5, 12, 22, 35, 51, 70, 92, 117, 145]
    """
    p = 0
    yield p
    for i in range(1, 3 * n + 1, 3):
        p += i
        yield p


def get_pentagonal(n: int) -> int:
    """Get `n`-th pentagonal number

    Args:
        n: Index of pentagonal number

    Examples:
        >>> print(get_pentagonal(10))
        145
    """
    return n * (3 * n - 1) // 2


def pentagonal_index(n: int) -> int:
    """Get index of `n` in pentagonal sequence

    Returns:
        Index of pentagonal number n. If n isn't pentagonal number, returns index of closest pentagonal number below n

    Examples:
        `n` is pentagonal number

        >>> print(pentagonal_index(145))
        10

        `n` isn't triangular number

        >>> print(pentagonal_index(144))
        9
    """
    return (isqrt(24 * n + 1) + 1) // 6


def is_pentagonal(n: int) -> bool:
    """Check if number is pentagonal

    Returns:
        True if number is pentagonal, False otherwise

    Examples:
        >>> is_pentagonal(145)
        True
        >>> is_pentagonal(144)
        False
    """
    return sqrt(24 * n + 1) % 6 == 5


def hexagonal_numbers(n: int) -> Iterator[int]:
    """Generate `n` first numbers in hexagonal numbers sequence `OEIS A000384 <https://oeis.org/A000384>`_

    Args:
        n: Number of hexagonal numbers

    Yields:
        The next hexagonal number in the range of 0 to `n`

    Examples:
        >>> print(list(hexagonal_numbers(10)))
        [0, 1, 6, 15, 28, 45, 66, 91, 120, 153, 190]
    """
    p = 0
    yield p
    for i in range(1, 4 * n + 1, 4):
        p += i
        yield p


def get_hexagonal(n: int) -> int:
    """Get `n`-th hexagonal number

    Args:
        n: Index of hexagonal number

    Examples:
        >>> print(get_hexagonal(10))
        190
    """
    return n * (2 * n - 1)


def hexagonal_index(n: int) -> int:
    """Get index of `n` in hexagonal sequence

    Returns:
        Index of hexagonal number n. If n isn't hexagonal number, returns index of closest hexagonal number below n

    Examples:
        `n` is hexagonal number

        >>> print(hexagonal_index(190))
        10

        `n` isn't hexagonal number

        >>> print(hexagonal_index(189))
        9
    """
    return (isqrt(8 * n + 1) + 1) // 4


def is_hexagonal(n: int) -> bool:
    """Check if number is hexagonal

    Returns:
        True if number is hexagonal, False otherwise

    Examples:
        >>> is_hexagonal(190)
        True
        >>> is_hexagonal(189)
        False
    """
    return sqrt(8 * n + 1) % 4 == 3


def champernowne_digits(n: int) -> int:
    """Get number of digits formed by `n`-digit numbers in Champernowne constant

    Examples:
        >>> print(champernowne_digits(3))
        2889
    """
    return (10 ** n * (9 * n - 1) + 1) // 9

File: euler/R/calculus.R
source("euler/R/big_int.R")

#' Get sum of first n terms of arithmetic series
#' 
#' @param a First number in series
#' @param b Common difference of series
#' @param n Number of terms in series
#' @examples
#' sum_arithmetic_series(1, 2, 3)
#' sum_arithmetic_series(0.5, 0.75, 3)
sum_arithmetic_series <- function(a, d, n) (n * (2 * a + d * (n - 1)) / 2)

#' Get sum of n first terms of geometric series
#'
#' @param a First number in series
#' @param b Common ratio of series
#' @param n Number of terms in series
#' @examples
#' sum_geometric_series(1, 2, 3)
#' sum_geometric_series(0.5, 0.75, 3)
sum_geometric_series <- function(a, r, n) a * (r ^ n - 1) / (r - 1)

#' Generate n first Fibonacci numbers \link[OEIS A000045]{https://oeis.org/A000045}
#'
#' @param n Number of Fibonacci numbers
#' @param t Function to convert numbers
#' @examples
#' fibonacci_numbers(10)
#' fibonacci_numbers(10, as.integer)
fibonacci_numbers <- function(n, t = as.bigint) {
    f_prev <- t(1)
    f_curr <- t(1)
    result <- c(f_prev, f_curr)
    for (i in seq_len(n - 2)) {
        f_temp <- f_prev + f_curr
        f_prev <- f_curr
        f_curr <- f_temp
        result <- c(result, f_curr)
    }
    return(result)
}

#' Get n-th Fibonacci number
#'
#' @param n Index of Fibonacci number
#' @examples
#' get_fibonacci(10)
get_fibonacci <- function(n) ((1 + sqrt(5)) / 2) ^ n / sqrt(5) + 0.5

#' Get index of n in Fibonacci sequence
#'
#' @returns Index of Fibonacci number n. If n isn't Fibonacci number, returns index of closest Fibonacci number below n
#' @examples
#' fibonacci_index(55)
#' fibonacci_index(54)
fibonacci_index <- function(n) as.integer(log(n * sqrt(5) + 0.5, (sqrt(5) + 1) / 2))

#' Get length of Collatz chain
#'
#' @param n Starting number
#' @examples
#' collatz_chain <- len_collatz_chain()
#' collatz_chain(13)
len_collatz_chain <- function() {
    hashmap <- 1
    memoize <- function(n) {
        if (is.na(hashmap[n])) {
            if (n %% 2 == 0) hashmap[n] <<- 2 + memoize((3 * n + 1) / 2) else hashmap[n] <<- 1 + memoize(n / 2)
        }
        return(hashmap[n])
    }
    return(memoize)
}

#' Generate n first numbers in triangular numbers sequence \link[OEIS A000217]{https://oeis.org/A000217}
#'
#' @param n Number of triangular numbers
#' @examples
#' triangular_numbers(10)
triangular_numbers <- function(n) {
    result <- 1
    for (i in 2:n) result[i] <- result[i - 1] + i
    return(result)
}

#' Get n-th triangular number
#'
#' @param n Index of triangular number
#' @examples
#' get_triangular(10)
get_triangular <- function(n) n * (n + 1) / 2

#' Get index of n in triangular sequence
#'
#' @returns Index of triangular number n. If n isn't triangular number, returns index of closest triangular number below n
#' @examples
#' triangular_index(55)
#' triangular_index(54)
triangular_index <- function(n) (sqrt(8 * n + 1) - 1) %/% 2

#' Check if number is triangular
#'
#' @returns TRUE if number is triangular, FALSE otherwise
#' @examples
#' is.triangular(55)
#' is.triangular(54)
is.triangular <- function(n) sqrt(8 * n + 1) %% 2 == 1

#' Generate n first numbers in square numbers sequence \link[OEIS A000290]{https://oeis.org/A000290}
#'
#' @param n Number of square numbers
#' @examples
#' square_numbers(10)
square_numbers <- function(n) (1:n) ^ 2

#' Get n-th square number
#'
#' @param n Index of square number
#' @examples
#' get_square(10)
get_square <- function(n) n * n

#' Get index of n in square sequence
#'
#' @returns Index of square number n. If n isn't square number, returns index of closest square number below n
#' @examples
#' square_index(100)
#' square_index(99)
square_index <- function(n) as.integer(sqrt(n))

#' Check if number is square
#'
#' @returns TRUE if number is square, FALSE otherwise
#' @examples
#' is.square(100)
#' is.square(99)
is.square <- function(n) sqrt(n) %% 1 == 0

#' Generate n first numbers in pentagonal numbers sequence \link[OEIS A000326]{https://oeis.org/A000326}
#'
#' @param n Number of pentagonal numbers
#' @examples
#' pentagonal_numbers(10)
pentagonal_numbers <- function(n) {
    result <- 1
    for (i in seq(4, 3 * n, 3)) result <- c(result, tail(result, 1) + i)
    return(result)
}

#' Get n-th pentagonal number
#'
#' @param n Index of pentagonal number
#' @examples
#' get_pentagonal(10)
get_pentagonal <- function(n) n * (3 * n - 1) / 2

#' Get index of n in pentagonal sequence
#'
#' @returns Index of pentagonal number n. If n isn't pentagonal number, returns index of closest pentagonal number below n
#' @examples
#' pentagonal_index(145)
#' pentagonal_index(144)
pentagonal_index <- function(n) (sqrt(24 * n + 1) + 1) %/% 6

#' Check if number is pentagonal
#'
#' @returns TRUE if number is pentagonal, FALSE otherwise
#' @examples
#' is.pentagonal(145)
#' is.pentagonal(144)
is.pentagonal <- function(n) sqrt(24 * n + 1) %% 6 == 5

#' Generate n first numbers in hexagonal numbers sequence \link[OEIS A000384]{https://oeis.org/A000384}
#'
#' @param n Number of hexagonal numbers
#' @examples
#' hexagonal_numbers(10)
hexagonal_numbers <- function(n) {
    result <- 1
    for (i in seq(5, 4 * n, 4)) result <- c(result, tail(result, 1) + i)
    return(result)
}

#' Get n-th hexagonal number
#'
#' @param n Index of hexagonal number
#' @examples
#' get_hexagonal(10)
get_hexagonal <- function(n) n * (2 * n - 1)

#' Get index of n in hexagonal sequence
#'
#' @returns Index of hexagonal number n. If n isn't hexagonal number, returns index of closest hexagonal number below n
#' @examples
#' hexagonal_index(190)
#' hexagonal_index(189)
hexagonal_index <- function(n) (sqrt(8 * n + 1) + 1) %/% 4

#' Check if number is hexagonal
#'
#' @returns TRUE if number is hexagonal, FALSE otherwise
#' @examples
#' is.hexagonal(145)
#' is.hexagonal(144)
is.hexagonal <- function(n) sqrt(8 * n + 1) %% 4 == 3

#' Get number of digits formed by n-digit numbers in Champernowne constant
#'
#' @examples
#' champernowne_digits(3)
champernowne_digits <- function(n) (10 ^ n * (9 * n - 1) + 1) / 9